MemoryMappedFile support on Windows

ntherning opened the pull request:
This implements the icalls in file-mmap-windows.c used by MemoryMappedFile. The implementation has been inspired by the CoreFX MemoryMappedFile implementation.

This patch also fixes a few incompatibilities between .NET and Mono:
- The default MemoryMappedFileOptions has been changed from DelayAllocatePages
  to None.
- mono_pagesize() on Windows now return SYSTEM_INFO.dwPageSize, the same as
  .NET and CoreFX.
- MemoryMappedFileTest has been updated to succeed on .NET.
- A tests has been added to test correct DelayAllocatePages behavior (only
  supported on Mono on Windows).
- The posix MemoryMappedFile code has been changed to throw
  UnauthorizedAccessException rather than IOException when trying to create a
  view outside of the bounds of a MemoryMappedFile.
- Creating a named mapping larger than 4GB on 32-bit systems now throws
  ArgumentOutOfRangeException.

evincarofautumn commented on mcs/class/System.Core/Test/System.IO.MemoryMappedFiles/MemoryMappedFileTest.cs:
```diff
@@ -56,16 +56,28 @@ static String MkNamedMapping ()
+		public void FixtureSetUp ()
+		{
+			try {
+				// Try to cleanup from any previous NUnit run.
```
Shouldn’t this never happen? It seems like you could use `GetRandomFileName` or something.

ntherning replied:
We cannot use random file names as locked files cannot be cleaned up until the next time the test suite is run in a new process. And then we wouldn't know what to clean up.

evincarofautumn commented on mcs/class/System.Core/Test/System.IO.MemoryMappedFiles/MemoryMappedFileTest.cs:
```diff
@@ -77,9 +89,14 @@ static String MkNamedMapping ()
+				// This throws an exception under MS.NET and Mono on Windows,
+				// since the directory contains open files.
 				Directory.Delete (tempDir, true);
+			} catch (Exception) {
```
Couldn’t you just close the files? Also, this should be `IOException` if you’re only interested in catching that specific error.

ntherning replied:
Yeah, making sure everything is closed at the end of every test would be ideal. But it's hard to make sure that's the case when a test fails for some reason. And to cope with that we need to make sure every test uses new temp files otherwise the following tests will fail on Windows due to the file already being open and thus being locked from deletion. We do this in other tests already (e.g. AssemblyTest) to work around the problem with locked files on Windows.

evincarofautumn commented on mono/metadata/file-mmap-windows.c:
```diff
@@ -1,71 +1,413 @@
+	size_t length;
+} MmapInstance;
+
+enum {
```
This could be shared among these implementations.

ntherning replied:
The `MmapInstance` in `file-mmap-posix.c` is different. It has a field which isn't used in Windows which is why I didn't think it should be shared.

evincarofautumn commented on mono/metadata/file-mmap-windows.c:
```diff
@@ -1,71 +1,413 @@
 
-void
-mono_mmap_flush (void *mmap_handle)
+static void *open_handle (void *handle, MonoString *mapName, int mode, gint64 *capacity, int access, int options, int *error)
```
These `int *error` could be replaced with `IOError *error` or something like that, if you named the `enum`.

ntherning replied:
True, that would make it more readable. I just copied the style used in the `file-mmap-posix.c` file.

evincarofautumn commented on mono/metadata/file-mmap-windows.c:
```diff
@@ -1,71 +1,413 @@
+	CloseHandle ((HANDLE) mmap_handle);
+}
+
+void mono_mmap_configure_inheritability (void *mmap_handle, gboolean inheritability)
```
Might be more clearly named `mono_mmap_set_inheritable` if this is all it does.

evincarofautumn commented on mono/metadata/file-mmap-windows.c:
```diff
@@ -1,71 +1,413 @@
-void *
-mono_mmap_open_file (MonoString *path, int mode, MonoString *mapName, gint64 *capacity, int access, int options, int *error)
+// These control the retry behaviour when lock violation errors occur during Flush:
+#define MAX_FLUSH_WAIT 15  // must be <=30
```
This would be clearer as `MAX_FLUSH_RETRIES`. (It’s a number of retries, not an amount of time to wait.)

ntherning replied:
Agreed. The retry behavior was ported from the CoreFX C# code where it´s called `MaxFlushWaits`. Looks like  the last `s` got lost. I'll change it.

evincarofautumn commented on mono/metadata/file-mmap-windows.c:
```diff
@@ -1,71 +1,413 @@
+	// TODO: Propagate error to caller
+}
+
+int mono_mmap_map (void *handle, gint64 offset, gint64 *size, int access, void **mmap_handle, void **base_address)
```
It seems like `*size` should be unsigned. Maybe also `offset`, although `off_t` is signed.

ntherning replied:
This is an icall which uses `long` in C#. The icall on the C# side is called by public APIs which also use `long` for size. The size is validated on the managed size to be positive so we could change the internal C# methods to take `ulong`. But I'm not sure it's worth the hassle to change the the icall signatures.

migueldeicaza commented:
I believe the changes to `mono_pagesize` are correct, reading the documentation for `VirtualAlloc` and `GetSystemInfo`, our original interpretation of those values was slightly erroneous.   The new usage is correct.

migueldeicaza commented:
On the rest of the changes: the shared bits are ok, and the Windows bits are entirely owned by Niklas team.   This patch can go in whenever Niklas wants to.

ntherning commented:
Thanks Miguel! I will wait for the PR builds before we merge this one.

kumpera commented:
The change to mono_pagesize is wrong. It must return the allocation granule
for mmap. That's the expected behavior by the runtime.

If you want, rename the existing function to say it returns the mmap
granule and add one that returns the actual min page size.

On Thursday, October 20, 2016, Niklas Therning <notifications@github.com>
wrote:

> This implements the icalls in file-mmap-windows.c used by
> MemoryMappedFile. The implementation has been inspired by the CoreFX
> MemoryMappedFile implementation.
>
> This patch also fixes a few incompatibilities between .NET and Mono:
>
>    - The default MemoryMappedFileOptions has been changed from
>    DelayAllocatePages to None.
>    - mono_pagesize() on Windows now return SYSTEM_INFO.dwPageSize, the
>    same as .NET and CoreFX.
>    - MemoryMappedFileTest has been updated to succeed on .NET.
>    - A tests has been added to test correct DelayAllocatePages behavior
>    (only supported on Mono on Windows).
>    - The posix MemoryMappedFile code has been changed to throw
>    UnauthorizedAccessException rather than IOException when trying to create a
>    view outside of the bounds of a MemoryMappedFile.
>    - Creating a named mapping larger than 4GB on 32-bit systems now
>    throws ArgumentOutOfRangeException.
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/mono/mono/pull/3796
> Commit Summary
>
>    - MemoryMappedFile support on Windows
>
> File Changes
>
>    - *M* mcs/class/System.Core/Microsoft.Win32.SafeHandles/
>    SafeMemoryMappedViewHandle.cs
>    <https://github.com/mono/mono/pull/3796/files#diff-0> (4)
>    - *M* mcs/class/System.Core/System.IO.MemoryMappedFiles/
>    MemoryMappedFile.cs
>    <https://github.com/mono/mono/pull/3796/files#diff-1> (20)
>    - *M* mcs/class/System.Core/System.IO.MemoryMappedFiles/
>    MemoryMappedView.cs
>    <https://github.com/mono/mono/pull/3796/files#diff-2> (2)
>    - *M* mcs/class/System.Core/Test/System.IO.MemoryMappedFiles/
>    MemoryMappedFileTest.cs
>    <https://github.com/mono/mono/pull/3796/files#diff-3> (96)
>    - *M* mono/metadata/file-mmap-posix.c
>    <https://github.com/mono/mono/pull/3796/files#diff-4> (20)
>    - *M* mono/metadata/file-mmap-windows.c
>    <https://github.com/mono/mono/pull/3796/files#diff-5> (408)
>    - *M* mono/utils/mono-mmap-windows.c
>    <https://github.com/mono/mono/pull/3796/files#diff-6> (2)
>
> Patch Links:
>
>    - https://github.com/mono/mono/pull/3796.patch
>    - https://github.com/mono/mono/pull/3796.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mono/mono/pull/3796>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAWav7rqTkh0Bk5hp1lSeFPtertGahOks5q12GQgaJpZM4KcFEE>
> .
>

ntherning commented:
Thanks @kumpera  for taking a look!

`Environment.SystemPageSize` should return 4k on Windows x64. The .NET docs says:

    In Windows, this value is the dwPageSize member in the SYSTEM_INFO structure.

so the currently returned value is wrong and unexpected.  It's currently an icall directly to `mono_pagesize()`. We either need to rewire the icall or return `dwPageSize` on Windows. I agree that adding a new function which returns the granule would be a good solution. I have added a `mono_valloc_granule()` function to the PR which returns `SYSTEM_INFO.dwAllocationGranularity` on Windows and on other platforms it just returns the page size from `mono_pagesize()`. 

Would you mind pointing me at any code which currently use `mono_pagesize()` but should call this new function? I have looked through all calls to `VirtualAlloc()` and AFAICS there's only one code path that actually passes a non-NULL address to `VirtualAlloc()` (`mono-codeman.c` calls `mono_valloc()` with a non-NULL address). Are you aware of any other Windows APIs which might be sensitive to the allocation granularity? Any other places in the runtime code I should look at?

If I understand the code in `mono-codeman.c` it will call `mono_valloc()` in `chunk_size` chunks and tries to allocate adjacent regions. Minimum chunk size is currently

```c
mono_pagesize() * MIN_PAGES
```

I have modified the code in `mono-codeman.c` to use

```c
MAX(mono_pagesize() * MIN_PAGES, mono_valloc_granule())
```

as minimum chunk size and also align on granule rather than page size. Since the granule and page size are equal on other platforms this should only affect Windows.

Comments?

kumpera commented:
LGTM.
The enums on the C side should be cleaned up though.

ntherning commented:
@kumpera Thanks for merging! I will fix that enum, move it to a shared header file.
